pandas/core/arrays/datetimes.py

```diff
@@ -118,27 +118,36 @@ def wrapper(self, other):
         else:
             if isinstance(other, list):
                 # FIXME: This can break for object-dtype with mixed types
-                other = type(self)(other)
-            elif not isinstance(other, (np.ndarray, ABCIndexClass, ABCSeries)):
+                try:
+                    other = type(self)(other)
+                except ValueError:
+                    other = np.array(other, dtype=np.object_)
+            elif not isinstance(other, (np.ndarray, ABCIndexClass, ABCSeries,
+                                        DatetimeArrayMixin)):
                 # Following Timestamp convention, __eq__ is all-False
                 # and __ne__ is all True, others raise TypeError.
                 return ops.invalid_comparison(self, other, op)
 
             if is_object_dtype(other):
                 result = op(self.astype('O'), np.array(other))
+                o_mask = isna(other)
             elif not (is_datetime64_dtype(other) or
                       is_datetime64tz_dtype(other)):
                 # e.g. is_timedelta64_dtype(other)
                 return ops.invalid_comparison(self, other, op)
             else:
                 self._assert_tzawareness_compat(other)
-                result = meth(self, np.asarray(other))
+                if not hasattr(other, 'asi8'):
+                    # ndarray, Series
+                    other = type(self)(other)
+                result = meth(self, other)
+                o_mask = other._isnan
```

Comment:
`isna(other)` ?

Reply:
It turns out that `isna(DatetimeArray)` doesn't work at the moment.  I want to keep the footprint of these PRs contained as long as possible, so chose to avoid changing it for now.

Reply:
But you are adding the `isna` method in this PR (to DatetimelikeArrayMixin above), shouldn't it then work?

Reply:
`isna(other)` may not work until `DatetimelikeArray(mixin)` inherits from ExtensionArray. I suspect there's a check in `isna` that only dispatches for EAs.

Reply:
> isna(other) may not work until `DatetimelikeArray(mixin)` inherits from ExtensionArray. I suspect there's a check in `isna` that only dispatches for EAs.

Generally right.  The check also catches a couple other things.